src/runtime/hexagon/hexagon_wrapper_dl.cc

```diff
@@ -0,0 +1,150 @@
+/*!
+ * Copyright 2019 Qualcomm Innovation Center, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE‐2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * This Contribution is being provided by Qualcomm Technologies, Inc.,
+ * a Delaware corporation, or its subsidiary Qualcomm Innovation Center, Inc.,
+ * a California corporation, under certain additional terms and conditions
+ * pursuant to Section 5 of the Apache 2.0 license.  In this regard, with
+ * respect to this Contribution, the term "Work" in Section 1 of the
+ * Apache 2.0 license means only the specific subdirectory within the TVM repo
+ * (currently at https://github.com/dmlc/tvm) to which this Contribution is
+ * made.
+ * In any case, this submission is "Not a Contribution" with respect to its
+ * permitted use with any of the "vta" and "verilog" subdirectories in the TVM
+ * repo.
+ */
+#include "hexagon_wrapper_dl.h"
+
+#include <assert.h>
+#include <stdlib.h>
+
+#include <iostream>
+
+#if defined(_WIN32)
+#include <windows.h>
+#else
+#include <dlfcn.h>
+#endif
+
+#ifndef HEXAGON_TOOLCHAIN
+#error HEXAGON_TOOLCHAIN macro must be defined
+#endif
+
+void* HexagonWrapperDL::resolve(const char* sym_name) {
+  assert(sim_dl_handle != nullptr && "libwrapper.so must be loaded");
+  dlerror();
+  if (void* sym = dlsym(sim_dl_handle, sym_name)) return sym;
+  fprintf(stderr, "dlsym failed for: %s, error: %s\n", sym_name, dlerror());
+  abort();
+}
+
+HexagonWrapperDL::HexagonWrapperDL(HEXAPI_Cpu cpu, HEX_4u_t flags,
+                                   HEX_4u_t recordflag) {
+  sim_dl_handle =
+      dlopen(HEXAGON_TOOLCHAIN "/lib/iss/libwrapper.so", RTLD_LAZY);
+  assert(sim_dl_handle);
+  sim_wrapper = malloc(sizeof(HexagonWrapper));
+  assert(sim_wrapper != nullptr);
+
+  void* sym_ctor = resolve("_ZN14HexagonWrapperC1E13hexagon_cpu_tjj");
```

Comment:
Not sure if it is good to hardcode the path and symbols

Reply:
Seems it only used on Linux Simulator and will not run Android target. However, Hexagon SDK has Windows version, so it maybe has problem on Windows Hexagon SDK. Need @kparzysz clarify.

Reply:
The library libwrapper.so is the shared object that implements the simulator functionality.  (It actually loads other .so files that implement specific architectures, but that's not that important.)  It's an x86 object and won't load on Android, so libwrapper.so cannot be automatically loaded (via the dynamic section of the ELF file), instead we load it conditionally via dlopen.  Since libwrapper.so is written in C++, the hardcoded symbols are mangled symbol names.

We could reimplement this by creating another wrapper on top of libwrapper.so: this extra wrapper would implement the functions from libwrapper.so through functions with C linkage.  Then, dlsym would have pretty names in it, instead of the mangled ones, but I'm not sure if this is worth the effort, as it would require another target in the Makefile structure.

Reply:
Hexagon SDK for Linux has a Linux toolchain that has the simulator running on Linux as well.